pennylane/circuit_graph.py

```diff
@@ -26,13 +26,27 @@ def _by_idx(x):
     return x.queue_idx
 
 
+def _is_observable(x):
+    """Predicate for deciding if an Operation instance is an observable.
+
+    Note that some :class:`Observable` instances are not observables in this sense,
+    since they can be used as gates as well.
+
+    Args:
+        x (Operation): node in the circuit graph
+    Returns:
+        bool: True iff x is an observable
+    """
+    return getattr(x, 'return_type', None) is not None
```

Comment:
Possibly, but I necessarily wouldn't. The predicate defines what "observable" means in the context of this module, even if it's just a trivial `isinstance(x, Observable)`. Also, if we later introduce other predicates for other purposes it's easy to compare them and see how they differ.

Reply:
BTW, it seems that right now we still have to keep this definition, since there are Observable subclasses such as PauliX which can also represent gates. Are there plans to separate the two meanings somehow?

Reply:
I think a cleaner solution would be that an Observable instance is always is an observable, and the PauliX gate is handled somehow differently.

Reply:
We'll have to keep thinking about this, but I suppose the only requirement is that the user doesn't have to use two separate `PauliZ` objects depending if it's an observable or an operation

Reply:
Why not? IMO a PauliZ gate and a PauliZ observable are conceptually two different things, they just happen to be represented by the same matrix.

Reply:
I would disagree. Of course, gates and observables are two different things, but the PauliZ is always the same and it would make no sense to me as a user to have `PauliZ` and `PauliZObservable` or whatever.

I think that having a predicate rooted in the return_type attribute is actually not that bad.

Reply:
I guess what I'm trying to say is that the fact that PauliZ the gate and PauliZ the observable are represented by the same matrix is coincidental in a sense. Gates are unitary, observables hermitian, this matrix happens to be both. I'm generally in favor of an operational naming of things, in which a gate and an observable ought to be different. Note that the difference could be just a different namespace, e.g. `gate.PauliZ` vs. `obs.PauliZ`.

That said, the current method will work too (at the price of slightly more complicated code), and we can use the predicate `_is_observable` as the *definition* of a quantum mechanical observable.

Reply:
I totally understand your train of thought, but I still disagree in a sense. The difference only lies in _how_ you use it (as an operation or as an observable) and not in the object itself, which is both operation and observable at the same time. So, operationally speaking it is different, yes. But object-wise it is exactly the same thing. 

Think of a cup. You can (operationally) use the same cup to drink tea and coffee (probably also other things). With the operational logic you would have to have a `CoffeeCup` and a `TeaCup`. But in the object-oriented sense you only need the `Cup` and that makes much more sense to me and would also make more sense to a user.